[WebProfilerBundle] Update the mailer panel

javiereguiluz opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Related to #47148, this updates one of the panels which weren't updated in that other PR.

### Before - 1 email sent

![before-1-email](https://user-images.githubusercontent.com/73419/191480458-a49d6f21-4119-47a7-9bb6-79f583e5cdc3.jpg)

### Before - Multiple emails sent

![before-multiple-emails](https://user-images.githubusercontent.com/73419/191480487-9c177e2c-e340-4b67-b0e3-9cc0fd6e4256.jpg)

### Before - Email attachments

![before-attachment](https://user-images.githubusercontent.com/73419/191480510-597b4e3c-8846-4ad3-b2c4-d8600abc15d6.jpg)

-----

### After - 1 email sent

![after-1-email](https://user-images.githubusercontent.com/73419/191480683-8849eeae-e034-414b-aa44-c6a9f25684c9.jpg)

Comments:

* All email contents are displayed on the same place, to make debugging quicker
* All headers are displayed too; this can be long in some cases, but I think it's better to display them all to spot errors easier and quicker
* Attachments now display file name, file size and a link to download them as files. We no longer display the base64-encoded contents of the file

### After - Multiple emails sent

![after-multiple-emails](https://user-images.githubusercontent.com/73419/191481013-c245f0d4-4702-49c6-b723-16208dbce20e.jpg)

Comments:

* When there's more than 1 email sent/queued, we display the "Email 1", "Email 2", etc. navigation (which is hidden when there's only 1 email to make design more efficient)

### After - MIME parts

![after-mime-parts](https://user-images.githubusercontent.com/73419/191481239-7e056578-17d1-42ba-91e9-e734fcf92f14.jpg)

Comments:

* This is the same as before

### After - Raw message

![after-raw-message](https://user-images.githubusercontent.com/73419/191481300-3275b0de-0108-4195-a1b5-19a8986c762f.jpg)

Comments:

* We now include a link to download the raw email as a `*.eml` file

javiereguiluz committed:
[WebProfilerBundle] Update the mailer panel

stof commented:
One thing that seems gone is showing which transport was used for each email (previously, it was shown in the tabs switching between emails)

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/mailer.html.twig:
```diff
@@ -101,135 +63,176 @@
+            </div>
+        {% else %}
+            {% set email_message = (collector.events.events(transport)|first).message %}
+            {{ _self.render_email_details(collector, collector.events.events(transport), email_message) }}
```
the second argument should be the event, not the list of events

javiereguiluz replied:
I've removed that second argument because it was not needed. Thanks.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/mailer.html.twig:
```diff
@@ -101,135 +63,176 @@
+        {%- if bytes < 1024 -%}
+            {{- bytes ~ ' bytes' -}}
+        {%- elseif bytes < 1024 ** 2 -%}
+            {{- (bytes / 1024)|number_format(2) ~ ' KB' -}}
```
should be `kB` not `KB` (or even better `kiB and MiB` as we use power of 2 and not power of 10)

javiereguiluz replied:
Updated to use `kB` and `1000`. Thanks!

javiereguiluz commented:
@stof thanks for the review. About the transport, to reduce the noise I thought it was a good idea to only display the transport if there's more than one transport. This is how it'd look like when showing the transport name:

<img width="1090" alt="mailer-transport" src="https://user-images.githubusercontent.com/73419/191509028-9fdbd895-eebb-415d-9c3c-a72ef62ac96f.png">

But thanks to your comment, I realized that maybe your app defines multiple transports, but this request only used one of them ... so you need the transport info somewhere. What about showing it like this:

<img width="742" alt="mailer-transport-information" src="https://user-images.githubusercontent.com/73419/191512144-58ac5483-2e16-477d-b0f7-230827ec3043.png">

Do you know how long can be the transport name in practice? It's the full DNS or a human-friendly short name? Thanks!

stof commented:
IIRC, it is the full DSN (minus the auth part)

javiereguiluz commented:
Given that it's the DSN, I've updated the design as follows:

<img width="632" alt="transport-sent" src="https://user-images.githubusercontent.com/73419/191517298-ae94ec09-9fe7-4f08-8f2a-57f7d70e65c5.png">

<img width="616" alt="transport-queued" src="https://user-images.githubusercontent.com/73419/191517313-c9fa88b8-583b-4682-8cb2-e5a83b17a972.png">

ro0NL commented:
what about showing multiple emails inbox-like? eg. a list of subject and date.

stof commented:
@ro0NL I doubt you will have web requests sending lots of emails (such case must be handled asynchronously). So it is better to optimize the experience for a few emails than for the case of a big list IMO.

ro0NL commented:
yeah but im not sure how not seeing the subject improves the experience ...

stof commented:
Well, if you want to display the subject instead of just `#1` and `#2`, it requires moving away from horizontal tabs then (as the subject won't fit in them), which means adding much more space for this email switcher.

94noni commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/mailer.html.twig:
```diff
@@ -101,135 +63,184 @@
+                    <div class="tab">
+                        <h3 class="tab-title">Raw Message</h3>
+                        <div class="tab-content">
+                            <a class="mailer-message-download-raw" href="data:application/octet-stream;base64,{{ collector.base64Encode(message.toString()) }}" download="email.eml">
```
small thing, if doable
perhaps add some unitary string data into the `email.eml` to have different names in case of multi email?

javiereguiluz replied:
I'd say we shouldn't use the email subject (because it can be too long; it can contain problematic characters for file names, which we cannot slugify without introducing other issues, etc.)

Given that all browsers auto-append a numeric suffix when downloading the same filename ... when having multiple emails you'll end up with `email.eml`, `email (1).eml`, `email (2).eml`, etc.

I think it's OK for this feature, which will be mostly used as: click to download email; double click downloaded file to open in an email client; check that it looks OK; delete the file.

javiereguiluz commented:
I liked @ro0NL proposal, so I gave it a shot ... and I think it works better. It now shows a clickable list of messages to follow the familiar behavior of email clients:

![mailer-email-list](https://user-images.githubusercontent.com/73419/191545703-4db85b42-a9aa-4301-8d22-0d438df844b0.gif)

We also protect the design against too long lists:

<img width="1108" alt="mailer-list-long" src="https://user-images.githubusercontent.com/73419/191545777-9363b838-7ae4-4769-9695-0944e6ff66f5.png">

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/mailer.html.twig:
```diff
@@ -73,165 +73,181 @@
+                        </thead>
+                        <tbody>
+                            {% for event in collector.events.events(transport) %}
+                                <tr class="mailer-email-summary-table-row {{ loop.first ? 'active' }}" data-target="#email-{{ loop.index }}">
```
to be accessible, we would need to have a `<button>` in the row, that can be used for keyboard navigation.

javiereguiluz replied:
I've decided to not add that ... because, sadly, the rest of the Symfony Profiler tool is not 100% accessible.

stof replied:
well, not being 100% accessible does not mean we should be even less accessible than before...

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -834,6 +834,20 @@ if (typeof Sfjs === 'undefined' || typeof Sfjs.loadToolbar === 'undefined') {
                 });
             },
 
+            initializeMailerTable: function() {
```
Woudln't it be possible to avoid having all the JS code specific to panels in base_js.html.twig, which is the JS code shared between the toolbar and the profiler UI ? This currently increases the size of the toolbar.

stof replied:
A solution is to have this code in the `<script>`  tag of the panel directly instead of having a script tag that calls this method.

stof replied:
btw, the same should probably be done for `updateLogsTable` and maybe other methods

javiereguiluz replied:
OK ... but let's do that in another PR so we can extract all foreign JS code at once. Thanks

javiereguiluz commented:
@stof I thought about what you said and I think you were right. That's why I've changed my mind and introduced some buttons in the table of emails to make it accessible.

This is how it looks when using a browser with accessibility capabilities. In regular browsers, the buttons are completely hidden:

<img width="991" alt="table-accessible" src="https://user-images.githubusercontent.com/73419/191679014-85918c2a-a3bd-44fd-83dc-f31fbd477c0e.png">

So, I think this feature is now ready and complete. Thanks.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -834,6 +834,33 @@ if (typeof Sfjs === 'undefined' || typeof Sfjs.loadToolbar === 'undefined') {
+                    });
+                };
+
+                emailRowButtons.forEach((emailButton) => {
```
As the button is inside the row, this code is useless. The click on the button will reach the listener on the row due to event bubbling.

javiereguiluz replied:
OK, I've simplified JS code too. Thanks!

94noni commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/mailer.html.twig:
```diff
@@ -101,135 +63,202 @@
+                                    <td>{{ loop.index }}</td>
+                                    <td>{{ event.message.headers.get('subject').bodyAsString() ?? '(No subject)' }}</td>
+                                    <td>{{ (event.message.headers.get('to').bodyAsString() ?? '(empty)')|replace({'To:': ''}) }}</td>
+                                    <td class="visually-hidden"><button class="mailer-email-summary-table-row-button" data-target="#email-{{ loop.index }}">View email details</button></td>
```
i don't know the feasibility but can it be swapped to use default html details https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details
(to avoid javascript)

stof replied:
this won't work. It  can only expand details inside the cell of the table. That's not what we want in this design.

fabpot commented:
Thank you @javiereguiluz.
